### ISSUE TITLE
Distributed tensorflow worker hangs at TF_CloseSession() when using MonitoredTrainingSession

### ISSUE THREAD

**Issue by leewyang** [tensorflow/tensorflow#21745]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: N/A
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:  RHEL7 and also Mac OS X 10.13.6
- **TensorFlow installed from (source or binary)**: binary (pip install)
- **TensorFlow version (use command below)**: 1.9.0+
- **Python version**: 2.7 (RHEL7), 3.6 (Mac)
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: See below

### Describe the problem
When using `MonitoredTrainingSession` in TensorFlow version 1.9 or higher, I'm seeing the following deadlock/hang (as reported by the `hanging-threads` pip package) when the context manager exits.  Note: I do not see this hang for versions 1.8 or earlier.  Also, note that this does not occur if using the older `tf.train.Supervisor` API.
```
----------     Thread 140682110711616 hangs       ----------
	File "trainer.py", line 102, in <module>
		tf.app.run(main=main, argv=[sys.argv[0]] + unparsed)
	File "/usr/lib/python2.7/site-packages/tensorflow/python/platform/app.py", line 125, in run
		_sys.exit(main(argv))
	File "trainer.py", line 67, in main
		print("step: {}".format(step))
	File "/usr/lib/python2.7/site-packages/tensorflow/python/training/monitored_session.py", line 689, in __exit__
		self._close_internal(exception_type)
	File "/usr/lib/python2.7/site-packages/tensorflow/python/training/monitored_session.py", line 726, in _close_internal
		self._sess.close()
	File "/usr/lib/python2.7/site-packages/tensorflow/python/training/monitored_session.py", line 974, in close
		self._sess.close()
	File "/usr/lib/python2.7/site-packages/tensorflow/python/training/monitored_session.py", line 1121, in close
		_WrappedSession.close(self)
	File "/usr/lib/python2.7/site-packages/tensorflow/python/training/monitored_session.py", line 974, in close
		self._sess.close()
	File "/usr/lib/python2.7/site-packages/tensorflow/python/training/monitored_session.py", line 974, in close
		self._sess.close()
	File "/usr/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 690, in close
		tf_session.TF_CloseSession(self._session)
```

The [code](https://gist.github.com/leewyang/7bf6d0df0328fbc4e2e97275d741a044) that generates this is based on the [Distributed Tensorflow](https://www.tensorflow.org/deploy/distributed) documentation (with a trivial/dummy model).  I start one PS node and two worker nodes on a single box as follows:
```
rm -rf /tmp/train_logs; \
python trainer.py \
     --ps_hosts=localhost:2222 \
     --worker_hosts=localhost:2223,localhost:2224 \
     --job_name=ps --task_index=0

python trainer.py \
     --ps_hosts=localhost:2222 \
     --worker_hosts=localhost:2223,localhost:2224 \
     --job_name=worker --task_index=0

python trainer.py \
     --ps_hosts=localhost:2222 \
     --worker_hosts=localhost:2223,localhost:2224 \
     --job_name=worker --task_index=1
```

I've been able to reproduce this quite consistently on:
- Mac 10.13.6, Python 3.6, TensorFlow 1.10
- RHEL7, Python2.7, TensorFlow 1.9

And the symptom goes away when switching to 1.8 or earlier.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by leewyang** [comment]
I've updated the original description w/ N/As.

**Comment by PigApple** [comment]
Any news for this defect, I got similar issue for distributed training.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by asimshankar** [comment]
@yuefengz : Mind taking a look?

(CC @mrry FYI)

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yuefengz: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by formath** [comment]
Same problem for me.

**Comment by ChenjunZou** [comment]
Any update 
I also met the problem.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yuefengz: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by yuzhichang** [comment]
I also met this problem.
OS Platform and Distribution (e.g., Linux Ubuntu 16.04): CentOS 7.6
TensorFlow installed from (source or binary): binary (pip install)
TensorFlow version (use command below): 1.12.0
Python version: 2.7.5

**Comment by zhjunqin** [comment]
same issue for me

distributed training on tf 1.10 , same code works on tf 1.7

master is hang on following traceback, but worker already closed

```
Stack for thread 140497678165760
  File "mnist.py", line 235, in <module>
    tf.app.run()
  File "/usr/local/lib/python3.6/site-packages/tensorflow/python/platform/app.py", line 125, in run
    _sys.exit(main(argv))
  File "mnist.py", line 232, in main
    sess.run(train_op)
  File "/usr/local/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 695, in __exit__
    self._close_internal(exception_type)
  File "/usr/local/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 732, in _close_internal
    self._sess.close()
  File "/usr/local/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 980, in close
    self._sess.close()
  File "/usr/local/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1127, in close
    _WrappedSession.close(self)
  File "/usr/local/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 980, in close
    self._sess.close()
  File "/usr/local/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 980, in close
    self._sess.close()
  File "/usr/local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 676, in close
    tf_session.TF_CloseSession(self._session)

```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yuefengz: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by lxl910915** [comment]
same issue for me too

**Comment by lxl910915** [comment]
in this function:
  master_impl_->CloseSession(request, response, [&n, &ret](const Status& s) {
    ret.Update(s);
    n.Notify();
  });
.
 But ret.Update(s) and n.Notify() are not called.

**Comment by erwa** [comment]
I had a similar issue doing asynchronous training with MNIST where once the first worker finishes, the rest of the workers would hang. I managed to solve this by disabling communication between workers (each worker only needs to talk to the ps): https://github.com/linkedin/TonY/pull/120/files

```
config_proto = tf.ConfigProto(device_filters = ['/job:ps', '/job:worker/task:%d' % task_index])
...
with tf.train.MonitoredTrainingSession(master=server.target,
                                       is_chief=(task_index == 0),
                                       checkpoint_dir=FLAGS.working_dir,
                                       hooks=hooks,
                                       config=config_proto) as sess:
...
```

**Comment by lxl910915** [comment]
@erwa @yuefengz 
The function Master::CloseSession in master.cc, the following code is blocked at session->Close()
	  SchedClosure([session, done]() {
		Status s = session->Close(); 
		session->Unref();
		done(s);
	  });

**Comment by lxl910915** [comment]
We set timeout for closing worker session.

**Comment by zjj2wry** [comment]
@erwa thank you, that works fine

**Comment by aie2** [comment]
I have a similar problem , when I use TPUEstimator.
In running "TPUEstimator.train",  I stop TPU instance, and it happens.

**Comment by Kublai-Jing** [comment]
Met the similar issue where the same code works fine for tf 1.6 and not for tf 1.12. Isn't this issue being solved ?

**Comment by mohantym** [comment]
Hi @leewyang !
It seems you are using older versions(1.x versions) of Tensorflow which is not supported any more. Could you confirm whether issue is replicating after [migrating](https://www.tensorflow.org/guide/migrate/migrate_tf2) to 2.6 or not ? Thanks!

**Comment by leewyang** [comment]
@mohantym yeah, this is really old now, and the TF2.x branch behaves differently, so feel free to close.

**Comment by mohantym** [comment]
Ok @leewyang ! Thanks for the confirmation.

**Comment by mohantym** [comment]
Are you satisfied with the resolution of your issue?
[Yes](https://goo.gl/forms/Oe0tEvODFRoI2gJF3)
[No](https://goo.gl/forms/fUjzOfrtkFbrOT8d2)
